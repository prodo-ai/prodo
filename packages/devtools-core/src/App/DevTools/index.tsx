import * as React from "react";
import styled from "styled-components";
import { dispatch, watch, state } from "../../model";
import { HeaderHeight, paddings, PanelWidth } from "../../styles";
import { Panel } from "../../types";
import { eventListener } from "../../utils/communication";
import { ActionLogPanel } from "./ActionLogPanel";
import { RenderLogPanel } from "./RenderLogPanel";
import { StatePanel } from "./StatePanel";

const panels: { [key in Panel]: React.ReactElement } = {
  state: <StatePanel />,
  actionLog: <ActionLogPanel />,
  renderLog: <RenderLogPanel />,
};

const StyledDevtools = styled.div`
  width: ${PanelWidth};
  min-height: 100vh;
  max-height: 100vh;

  display: flex;
  flex-direction: column;

  background-color: ${props => props.theme.colors.bg};
  color: ${props => props.theme.colors.fg};
`;

const Tabs = styled.div`
  height: ${HeaderHeight};

  display: flex;
  justify-content: space-around;

  border-bottom: 1px solid ${props => props.theme.colors.fg};

  cursor: pointer;
`;

const Tab = styled.div<{ isSelected: boolean }>`
  padding: ${paddings.small};
  ${props => props.isSelected && `font-weight: bold`};

  &:hover {
    color: ${props => props.theme.colors.accent};
  }
`;

const StyledPanel = styled.div`
  overflow: auto;
  padding: ${paddings.medium};
  min-height: calc(100vh - ${HeaderHeight});
  max-height: calc(100vh - ${HeaderHeight});
`;

const RecordButtons = styled.div`
  position: absolute;
  bottom: 5px;
  right: 5px;
`;

function post({ func, universe, actionLog }: any) {
  const name = prompt("Name:");
  // @ts-ignore
  const { port } = window.devtoolsServer;
  fetch(`http://localhost:${port}/${func}`, {
    method: "post",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ universe, actionLog, name }),
  })
    .then(res => res.json())
    .then(res => console.log(res));
}

const ServerActions = () => {
  // @ts-ignore
  const { buttons } = window.devtoolsServer;
  console.log({ ServerActions, buttons });
  const universe = watch(state.app.universe);
  const actionLog = watch(state.app.actionLog);
  return (
    <RecordButtons>
      {Object.keys(buttons).map(func => (
        <button onClick={() => post({ func, universe, actionLog })}>
          {buttons[func]}
        </button>
      ))}
    </RecordButtons>
  );
};

export const DevTools = () => {
  React.useEffect(() => {
    window.addEventListener("message", eventListener(dispatch));
    return () => window.removeEventListener("message", eventListener(dispatch));
  }, []);

  const [selectedPanel, setSelectedPanel] = React.useState("state" as Panel);
  // @ts-ignore
  const hasServer = !!window.devtoolsServer;
  console.log({ hasServer });

  // TODO: bring back scroll-to-bottom?
  return (
    <StyledDevtools className="devTools" data-testid="devTools">
      <Tabs className="headerTabs" data-testid="headerTabs">
        {Object.keys(panels).map(panel => (
          <Tab
            isSelected={selectedPanel === panel}
            onClick={() => setSelectedPanel(panel as Panel)}
            className={`${panel}HeaderTab`}
            data-testid={`${panel}HeaderTab`}
            key={panel}
          >
            {getPanelTitle(panel)}
          </Tab>
        ))}
      </Tabs>
      <StyledPanel>{panels[selectedPanel]}</StyledPanel>
      {hasServer && <ServerActions />}
    </StyledDevtools>
  );
};

// Title-cases and spaces panel name, eg. actionLog => Action Log
const getPanelTitle = (data: string) => {
  return (
    data.charAt(0).toUpperCase() +
    data
      .slice(1)
      .split("Log")
      .join(" Log")
  );
};
