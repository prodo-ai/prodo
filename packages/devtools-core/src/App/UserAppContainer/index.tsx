import * as React from "react";
import * as ReactDOM from "react-dom";
import styled from "styled-components";
import { dispatch, state, watch } from "../../model";
import { eventListener } from "../../utils/communication";
import ErrorBoundary from "./ErrorBoundary";

const StyledUserAppContainer = styled.div`
  flex-grow: 1;
  text-align: center;
`;

const StyledIFrame = styled.iframe`
  height: 100%;
  width: 100%;

  border: none;
`;

interface Props {
  children?: React.ReactNode;
}

let prevLoc = window.location.href;
let interval = -1;

const UserAppContainer = (props: Props) => {
  const iFrameRef = React.useRef<HTMLIFrameElement>(null);
  const [updateValue, forceUpdate] = React.useState(true);
  const [styleNodes] = React.useState([] as any[]);

  // Make sure styled components update on route change
  // If they use the route plugin, we can just watch the route path
  // (Sometimes it takes a little while to load the route)
  // If not, we have to poll the location to see if it changed
  const route = watch(state.app.universe.route.path);
  const checkLocation = () => {
    if (window.location.href !== prevLoc) {
      prevLoc = window.location.href;
      copyStyleNodes();
    }
  };
  if (!route && interval === -1) {
    interval = window.setInterval(checkLocation, 100);
  } else if (route && interval !== -1) {
    window.clearInterval(interval);
    interval = -1;
  }

  // Forces a re-render
  const handleLoad = () => {
    forceUpdate(!updateValue);
  };

  React.useEffect(() => {
    if (iFrameRef && iFrameRef.current && iFrameRef.current.contentWindow) {
      iFrameRef.current.contentWindow.addEventListener(
        "message",
        eventListener(dispatch),
      );

      // Needed for Chrome
      handleLoad();
      return () => {
        if (iFrameRef && iFrameRef.current && iFrameRef.current.contentWindow) {
          iFrameRef.current.contentWindow.removeEventListener(
            "message",
            eventListener(dispatch),
          );
        }
      };
    }
  }, []);

  React.useEffect(() => {
    copyStyleNodes();
  });

  const copyStyleNodes = () => {
    if (iFrameRef && iFrameRef.current && iFrameRef.current.contentDocument) {
      document.head.childNodes.forEach((link: ChildNode) => {
        if ((link as any).tagName === "STYLE") {
          const content = (link as any).innerHTML;
          // We use styled components too, so append a copy
          if (content.includes("sc-component-id:")) {
            const copy = link.cloneNode(true);
            styleNodes.push(copy);
          } else if (!content.includes("prodoDevtoolsStyles")) {
            styleNodes.push(link);
          }
        }
      });
      styleNodes.forEach(link => {
        iFrameRef.current!.contentDocument!.head.appendChild(link);
      });
    }
  };

  const renderFrameContents = () => {
    if (iFrameRef && iFrameRef.current && iFrameRef.current.contentDocument) {
      return [
        ReactDOM.createPortal(
          props.children,
          iFrameRef.current.contentDocument.body,
        ),
      ];
    }
    return [];
  };

  // onLoad={handleLoad} is needed for Firefox
  return (
    <StyledUserAppContainer
      className="userAppContainer"
      data-testid="userAppContainer"
    >
      <ErrorBoundary>
        <StyledIFrame
          ref={iFrameRef}
          className="iframe"
          data-testid="iframe"
          onLoad={handleLoad}
        >
          {renderFrameContents()}
        </StyledIFrame>
      </ErrorBoundary>
    </StyledUserAppContainer>
  );
};

export default UserAppContainer;
