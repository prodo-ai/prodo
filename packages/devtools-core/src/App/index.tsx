import * as React from "react";
import styled, { ThemeProvider } from "styled-components";
import darkTheme from "../styles/theme";
import { DevTools } from "./DevTools";
import UserAppContainer from "./UserAppContainer";

const Layout = styled.div`
  display: flex;
  justify-content: space-between;
`;

interface Props {
  children?: React.ReactNode;
  skipUserApp?: boolean;
}

export default (props: Props) => (
  <ThemeProvider theme={darkTheme}>
    <Layout className="devToolsApp" data-testid="devToolsApp">
      {!props.skipUserApp && (
        <UserAppContainer>{props.children}</UserAppContainer>
      )}
      <DevTools />
    </Layout>
  </ThemeProvider>
);
