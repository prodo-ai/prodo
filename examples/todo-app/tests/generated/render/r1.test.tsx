import * as React from "react";
import App from "../../../src/App";
import render from "../../utils/render";
describe("render", () => {
  it("test1", async () => {
    const { container } = render(<App />, {
      initState: { todos: { T1: { text: "milk", done: false } } },
    });
    expect(container).toMatchSnapshot();
  });
});
