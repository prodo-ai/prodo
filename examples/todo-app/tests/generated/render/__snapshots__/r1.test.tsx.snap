// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render test1 1`] = `
<div>
  <div
    class="app"
  >
    <h1
      class="sc-kgoBCf uiiTR title"
    >
      Todos
    </h1>
    <input
      aria-label="item-input"
      placeholder="What needs to be done?"
      type="text"
    />
    <ul
      data-testid="list"
    >
      <li
        data-testid="item"
      >
        <input
          type="checkbox"
        />
        <span
          class="item-text"
        >
          milk
        </span>
        <button>
          x
        </button>
      </li>
    </ul>
    <div
      class="buttons"
    >
      <button>
        delete all
      </button>
    </div>
  </div>
</div>
`;
